classes/local/factor/object_factor_base.php

```diff
@@ -35,6 +35,20 @@ abstract class object_factor_base implements object_factor {
      */
     public $name;
 
+    /**
+     * Factor priority.
+     *  This is a number from 1 to 3 that determines order the factor is being authenticated.
+     *
+     * 1 - high priority. Factors without login form should go with this priority.
+     *
+     * 2 - medium priority.
+     *
+     * 3 - low priority.
+     *
+     * @var int
+     */
+    public $priority;
```

Comment:
I may not have been clear in the spec, this isn't a priority set by the factor code, this should be an order set by the admin and configured. This should be similar to explicitly setting an order for the authentication plugins. Plugins should not have any say in when they fire. By default each plugin won't really have an order, when it is enabled then it will get added to the end of the list. You need to add UI to move them up and down, which is similar to the enable / disable code. Some places in core also allow drag and drop but it's overkill for here to don't bother.

Reply:
I was following this spec: "for simplicity this is declared by the plugin and cannot be changed".
https://github.com/catalyst/moodle-tool_mfa/issues/11
But, that's ok, it won't take much time to implement this way ^^

Reply:
Ah yes sorry. This has evolved quite a bit :/